cranelift-codegen/meta/src/isa/x86/encodings.rs

```diff
@@ -41,6 +45,32 @@ impl<'defs> PerCpuModeEncodings<'defs> {
         }
     }
 
+    /// Add an instruction encoding using the given pattern; the pattern allows the builder to be
+    /// re-used for adding the same encoding multiple times with variations, e.g. for adding REX
+    /// variants.
+    fn add(&mut self, pattern: EncodingPattern, builder: EncodingBuilder) {
+        let encoding = builder.build(&mut self.recipes, &mut self.inst_pred_reg);
+        match pattern {
+            EncodingPattern::X64 => self.enc64.push(encoding),
+        }
+    }
+
+    /// Create an encoding builder from a recipe.
+    fn enc_rec(
+        &self,
+        inst: impl Into<InstSpec>,
+        recipe: EncodingRecipe,
+        bits: u16,
+    ) -> EncodingBuilder {
+        EncodingBuilder::new(inst.into(), recipe.clone(), bits, self.formats)
+    }
+
+    /// Create an encoding builder from a template.
+    fn enc(&self, inst: impl Into<InstSpec>, template: Template) -> EncodingBuilder {
```

Comment:
Is the idea to remove the other helper functions, and just use a combination of these new ones instead?

Reply:
Yes! For me it was initially very confusing to figure out what was going on in this file, if/should I create new helper methods, figuring out if a helper method I needed already existed, etc. I think having fewer methods (`add` and `enc`) would have helped me figure out more clearly what was going on.